stel/man/stel.1
.\" 
.\"  $Author: vince $
.\"  $Header: /users/vince/src/stel-dist/stel/RCS/stel.1,v 1.2 1996/03/26 14:42:29 vince Exp $
.\"  $Date: 1996/03/26 14:42:29 $
.\"  $Revision: 1.2 $
.\"  $Locker:  $
.\"
.TH stel 1
.ds )H CERT-IT
.ds ]W BETA Release: May 1996

.SH NAME

stel \- Secure TELnet
.SH SYNOPSIS
.B stel
[
.IR hostname
]  [
.IR options
]  [ 
.IR commands
] 

.SH DESCRIPTION

.IR stel(1)
is intented to be a
.IR simple
and
.IR practical
.IR "plug and play"
solution for logging into a remote system or executing
.IR commands
on a remote system.
.B stel
offers a variety of features which range from mutual
authentication to file transfer.  The security of
.B stel
relies upon the underlying cryptographic desing, not trusting
the net in any way.
.IR hostname
is the name of the remote system.
If no options are provided,
.B stel
defaults to triple DES encryption, 1024 bits Diffie Hellman
modulus, no protection from active attacks, and the remote
system user's shell as the command to be executed on the
remote system.  This should
be reasonable for most systems, and it is believed to be a
good compromise  between speed, ease of use and security.

.SH OPTIONS
.B stel
recognizes the following options:
.RS

.TP
.B -1
Use the
.IR "single DES"
cipher.  For the crypto impaired, single DES has a 64-bit block 
size for the plaintext and the ciphertext.  It uses a key size of 56
bits.   It is based on the design concept of combining substitutions
and permutations, and it has 16 rounds.
DES is probably the best studied encryption algorithm available today, in
fact it was first released in 1976.  Since then, there has been
much speculation on the key length, on the number of interations,
and the design of the S-boxes.  Some people even feared that the
NSA embedded a ``trap-door'' into the algorithm but, as far as it
is publicly known,
nobody has never proved or misproved anything.  The only real
.IR practical
problem DES is suffering from is key length.  This is why
.IR "triple DES"
remains the default encryption method.


.TP
.B -4
Use the
.IR RC4
cipher.  It is the (alleged) 
RC4 algorithm posted in the Usenet News in 1995 (RC4
was originally a trade secret algorithm of RSA Data Security and has
not been published; however, this implementation is widely believed to
be equivalent).
RC4 is a stream, variable-key-size cipher, and it uses a key size of
128 bits in the present implementation.  RC4 is a very new algorithm,
and it is not nearly so well-studied as DES or IDEA.  
RC4 is the fastest algorithm currently supported by
.CR stel ,
and it is intended to be used when speed is an issue and security 
is
.CR not .

.TP
.B -a
Protect from
.IR active
attacks.  This is one of the most advanced
features of
.CR stel .
When using this option, the data stream
which is transmitted from client to server and vice versa is
``packetized'' and sanity checks are performed on each
data packet.  Sanity checks include CRC32 encrypted MACs,
to foil garbage injection, and encrypted sequence numbers,
to foil replay attacks.  When using this option,
.IR "file transfer"
facilities are available.


.TP
.B -D
Enable
.IR debug
mode.  When this option is set, extensive debugging information
are displayed.  Most variables and keys are dumped, and they
are effectively visibile on to the screen, so use this option
with care.
.B -D
should only  be used to achieve a better
understanding about how
.B stel
really works.  When this option is set, both encrypted and unencrypted
data are dumped on to the screen, making it difficult to read what
is being displayed on the screen.

.TP
.B -e 
Disable
.IR escape
features.  By default,
.B stel
provides escapes features by means of the default escape
character ``^]''.  When this option is set, escapes
features are disabled.  This option is intended if
using
.B stel
in conjuction with modem file transfer programs
(i.e., kermit),
when a complete terminal transparency is required.

.TP
.B -h
Display some
.IR help
information.

.TP
.B -i
Use the
.IR IDEA
cipher.
For the crypto impaired, IDEA has a 64-bit block 
size for the plaintext and the ciphertext.  It uses a key size of 128
bits.  It is based on the design concept of ``mixing operations from
different algebraic groups''.  It has been developed at ETH in Zurich
by James L. Massey and Xuejia Lai, and published in 1990.
It is estimated that the workload to
exhaust all the possible 128-bit keys in the IDEA cipher would
roughly equal the factoring workload to crack a 3100-bit RSA key,
which is quite a bit bigger than the 1024-bit RSA key size that most
people use for high security applications.  IDEA is resistant to
differential and related-key cryptanalysis, and it could be
significantly more secure than DES.  However, it is not nearly
so well-studied as DES.  This is why triple DES remains the
default encryption method.

.TP
.B "-l username"
Set the user login name on the remote host to
.IR username
The default name is the current account name of the user invoking
.CR stel .

.TP
.B -m
Set Diffie-Hellman key exchange algorithm's
.IR modulo
to
.IR 512
bits.  Diffie-Hellman is used by
.B stel
for sharing a secret between client and server.  The DH algorithm
gets its security from the difficulty of calculating discrete
logarithms in a finite field, as compared with the ease of
calculating exponentiation in the same field.  512 bits are
considered the minimum size for achieving a minimum security
level.  
.B stel
uses, in fact, 1024 as default.  The size of the modulus
severely impacts the speed of the authentication process, so
this option is merely suggested when sitting at or connecting
to a very slow system.

.TP
.B -M
Set Diffie-Hellman key exchange algorithm's
.IR modulo
to
.IR 2048
bits.  2048 bits should be enough for the most security 
demanding contexts.  When using this option on slow
systems,
.B stel
timeouts before the authentication phase completes, so
use
.B -M
when when sitting at and connecting to fast systems.

.TP
.B -n
Do
.IR not
use data encryption at all.  This is intended to be a debugging
only option; when this option is set
.B stel
loses most of its
security strength.  Do not use it for real connections.

.TP
.B "-p port"
Use a non standard
.IR port
number.  If 
.B -p
is not used,
.B stel
attempts to contact a STEL server at the standard STEL  port
that is, 10005 or the one described by the 
.B stel
service.

.TP
.B -r
Enhance
.IR randomness .
When this option is set, the user is asked to
type in a reasonably long random string, which is passed, among with
other internal random information, through a strong hash
function (MD5), in order to distill randomness.  By default,
.B stel
generates internal random information
by means of system dependent information
(such as processes, frequently changing file's time stamps,
internal variables) and by means of the 
.IR truerand
package.

.TP
.B -t
Do not allocate a pseudo
.IR terminal
on the remote system.
Normally
.B stel
allocates a pseudo terminal on the remote system,
since standard shell sessions require it.
This option can be handy when executing
non-interactive commands on the remote system.  When this
option is set, it is not possible, for the STEL server,
to turn off printing while passwords are being typed in.
Use this option with care.

.TP
.B -v
Be
.IR verbose .
When this option is set, some debugging information, internal variables
and encryption keys are displayed.  Encryption session keys are
effectively visibile on to the screen, so use this option with care.

.RE

.SH USER AUTHENTICATION
Basically, 
.B stel 
supports three authentication methods:
.IR SecurID ,
.IR s/key
and
.IR "Unix passwords" .
The three systems are checked in order.
If the user is SecurID registered, by means of
.CR /etc/securid.conf ,
then SecurID authentication is required.  Else, if the user is
s/key registered, by means of
.B /etc/skeykeys 
or remotely, by means of the
.B skeyd 
s/key daemon and the 
.B /etc/skeydconf
configuration file, then s/key authentication is required.
Users can be forced not to use standard passwords, by means
of the 
.B /etc/skey.access 
configuration file.
Unix passwords are considered the last resort authentication
method.  However, being all data transmissions
encrypted, even standard Unix password offer a
reasonable level of security.
Finally, all logging-in users are checked against
.CR /etc/login.access ,
which controls access from the network.

.SH MUTUAL AUTHENTICATION

.B stel
uses the Diffie-Hellman exponential based method to determine
a common encryption key.  This completely eliminates the need of
keyservers to store and manage user keys and it greatly simplifies
the overall system design.  The Diffie-Hellman method does not
require any a-priori shared secret between the parties.  
The algorithm is based on the difficulty of factoring in
finite fields, and the protocol is very clever and very  simple.
However, it provides no authentication between the two parties.
A malicious hacker 
using an active line tap could intercept and change all messages,
impersonating the client to the server and the server to the
client.  The client and the server would not realize that;
for example, all messages sent by the client to the server
would be received by the hacker, decrypted by the hacker using
the client's session key, encrypted again using the server's
session key and sent to the server.  This is what is commonly called
.IR "Man In The Middle Attack" .

To defeat such an attack, a system based on the
.IR Interlock
Protocol (an idea by Shamir and Rivest)
is supported by 
.CR stel .
If the user wishing to login on the remote system owns
a file named 
.B .stelsecret
in his/her remote home directory
then the information contained in the file is exploited to
perform mutual authentication between the parties.
Similarly, system to system mutual authentication is
provided by means of
.CR /etc/stelsecret ,
which applies to all users of the system.
It is desiderable to keep the contents of
.B .stelsecret
and
.B /etc/stelsecret
files confidential, since mutual authentication depends on them.
It is up to the user to protect his/her
.CR .stelsecret ,
file while it is up to the adminitrator to protect
.CR /etc/stelsecret ,
eventually by making 
.B stel
a SUID program and chown-ing
.B /etc/stelsecret
to the UID
.B stel
is SUID to.
.B stel
attempts, in fact, to act safely in case it has been made
a SUID program.

The syntax of 
.B ~/.stelsecret
and
.B /etc/stelsecret
files is as follows:

.RS
.B "hostname1, hostname2, ... , hostN : secret-pass-phrase"
.RE 

.B secret-pass-phrase
is a single string (no spaces) of arbitrary length. 
Mutual authentication relies upon this secret;
maintenance of this secret, and distribution of its contents
to remote sites for use when connecting there is left as an
exercise for the reader.  The suggested way for doing this
is logging to the remote system's console and editing the
file.

.B hostname1
is a host name optionally expressed in dotted notation or
by means of regular expressions.  Supported regular
expressions are ``*'', ``?'' and ``[]''.  For example:

.RS
.B sun-[ab].some.domain.org
.RE

matches 
sun-a.some.domain.org
and 
system-b.some.domain.org ,
while

.RS
.B *.some.domain.org
.RE

matches all systems in the *.some.domain.org domain .
DOTS (``.'') are never matched by ``*'' or ``?'',
so *.some.domain.org does not match this.one.domain.org or
this.other.one.domain.org.

A sample, one line, stelsecret file may look like this:

.RS
.B work[123].domain.edu,10.20.30.4?,*.big.net:some-secret
.RE

When someone connects to a STEL server as
.IR username ,
the server checks
.B /etc/stelsecret
and
.IR username 's
.CR ~/.stelsecret
secret files on the server system.  If a match between the user's
source address and any host address contained in the secret files
occurs, the logging-in user is
required to type in the corresponding secret.  The typed-in
secret is verified by means of a strengthened version of the
.IR Interlock
Protocol
which is supposed to foil the
.IR "Man In The Middle Attack"
completely.


.SH ENVIRONMENT
The following environment variables are inherited by the server:

.RS
.CR TERM ,
.CR DISPLAY ,
.CR LINES ,
.CR COLUMNS ,
.CR WINDOWID
.RE

The following environment variables are set by the server,
upon the logging-in user being authenticated:

.RS
.CR LOGNAME ,
.CR USER ,
.CR USERNAME ,
.CR SHELL ,
.CR MAIL
.RE

The user is also provided with a new environment variable,
.CR STELKEYHASH ,
which is a one-way hashing of the actual session key.
The environment variable 
.CR SKEYPADFILE
is checked and it is used when generating s/key reponses
by means of the MJR DES-padding method.  Refer to
.CR MJR-SKEY-MODE
for details.

.SH ESCAPE FEATURES
The
.IR escape
menu is enabled by hitting the escape character once the
connection between client and server has been established.
Escape features include experimental ftp-like
.IR "file transfer"
commands (
.CR put ,
.CR get
),
a
.CR log
command, a
.CR status
command, and s/key
response generation commands (
.CR skey ,
.CR auto
). The
.CR auto
command detects s/key prompts and automatically
generates and sends s/key responses.
A special s/key operation mode,
.IR "MJR DES padding" ,
is available.  MJR DES padding mode is supposed to
foil s/key dictionary attacks.
Use of these commands should be straightforward.

.SH SECURITY OF STEL
The security of
.B stel
depends on different factor.  They are the strength of the
.IR "encryption algorithms"
, the strength of the
.IR Diffie-Hellman
key exchange method, the strength of the
.IR "random number generator" ,
the strength of the underlying
.IR protocol
, the correctness of
the implementation (that is,
.IR
.IR bugs !
) and, last but not least, the
.IR "Human Factor(tm)"
, which is universally known to be one of the
crucial factors for all security systems.

The
.IR "encryption algorithms"
used by
.B stel
are considered to be the state of the art algorithms available
today.
.IR RC4
is not intented for real secure communications, since
it is too new and untested, and rumors about alleged weak keys
families are enough for most security demanding users
to stay away from it.  On the
the other hand, RC4 is very fast, and it is handy when connecting
to or sitting at a very slow computer, or for file transfers.
.IR IDEA
is supposed to be very strong, and most cryptographers
are today betting on it.  But the algorithm is new, and there
could be some devastating cryptanalytic news about IDEA tomorrow.
.IR "Single DES"
is a well-known and well-built algorithm, and it is still
secure, unless you need your secrets to be protected from
the efforts of the major governments and organizations.
The default encryption method, that is
.IR "triple DES"
which three different pseudo-independent keys, is much
stronger than single DES.  The first
and third key are really independent and unrelated keys, and
the second key is generated using the IDEA cipher and the
other two keys.  Being considered the

rrc2/history
 ############################################################################
#                                                                            #
#                      History of my Implementation of                       #
#                      RRC.2 - Ron Rivest Cipher Nr. 2                       #
#                                                                            #
 ############################################################################

13.02.96:
	Received specification

15.02.96: 
	Linux (ANSI-) C version done;
	includes RRC.2 library, test program, and simple sample crypt
	program

18.02.96:
	First announcement on usenet (rrc2_news.txt)

18.02.96:
	Turbo Pascal version done;
	Includes only Unit RRC2 and test program.

20.02.96:
	Improved C version: now endian neutral (untested)

20.02.96:
	RRC.2 obtainable via ftp

23.02.96:
	second usenet announcement (rrc2_news_2.txt)

# version FINI


stel/stel/stel.c
/*
 * $Author: vince $
 * $Header: /users/vince/src/stel-dist/stel/RCS/stel.c,v 1.122 1996/04/26 16:29:23 vince Exp vince $
 * $Date: 1996/04/26 16:29:23 $
 * $Revision: 1.122 $
 * $Locker: vince $
 */

#ifndef	lint
static char *rcsid = "@(#) $Id: stel.c,v 1.122 1996/04/26 16:29:23 vince Exp vince $";
#endif	lint

#include <stdio.h>
#include <termios.h>
#include <signal.h>
#include <pwd.h>
#include <netinet/in.h>
#include <netdb.h>
#include <syslog.h>

#ifdef HAS_SYS_PARAM_H
#include <sys/param.h>
#endif

#ifdef AIX
#include <sys/ioctl.h>
#endif
#include "defs.h"

/* md5 stuff */
#include "global.h"
#include "md5.h"

/* idea stuff */
#include "idea.h"
extern IDEAkey		c_ideakey;	/* client */

/* des stuff */
/* #include "des.h" */
extern des_key_schedule	c_schedule1;	/* client */
extern des_key_schedule	c_schedule2;	/* client */
extern des_key_schedule	c_schedule3;	/* client */

/* IVs */
extern des_cblock	s_IV, c_IV;

extern int	logfh; /* log to file handler */

#define OPTIONS "l:np:tvrmMi14aDeh"

#define USAGE		\
"STEL: Secure TELnet, BETA VERSION, protocol id = %d\n\
Usage: %s [hostname] [options] [commands...]\n\
	hostname:	the system you want to connect to\n\
     	-l logname:	the username on the remote system\n\
	-p portnumber:	set port number\n\
	-i:		use IDEA cipher\n\
	-1:		use SINGLE DES cipher\n\
	-4:		use RC4 cipher\n\
	-n:		do not use data encryption\n\
	-r:		enter a random string to enhance randomness\n\
	-m:		use (small) 512 bits modulus\n\
	-M:		use (large) 2048 bits modulus\n\
	-a:		protect from active attacks\n\
	-t:		do not allocate a pseudo terminal\n\
	-e:		disable escape features\n\
	-v:		be verbose\n\
	-D:		be extra verbose\n\
	-h:		display this help screen\n\n\
Default port number is %d\n\
Default encryption algorithm is EDE TRIPLE DES\n\
Default Diffie-Hellman modulus size is 1024 bits\n\
Default escape character is '^]'\n"
#define USECRYPT	\
"This session is using %s encryption for all data transmissions.\r\n"
#define NOUSECRYPT	\
"WARNING: NOT USING DATA ENCRYPTION -- THIS SESSION IS INSECURE\r\n"
#define NOUSEPTY	\
"This session is not using pseudo terminals: passwords will be visible.\r\n"
#define BADOPTION	"option -%c is not compatible with previous options\n"
#define ACTIVEPROTECT	\
"This session is checking integrity and ordering for all data transmissions.\r\n"
struct termios		old, new;
int			ttyset;
static struct termios	saved_ttymode, noecho_ttymode; 

main(argc, argv)
int	argc;
char	*argv[];
{
	cpacket		pp;
	spacket		ppp;
	char		hostbuf[100], username[MAXUSERNAMELEN];
	char		*tohost, *env, *encmode;
	char		*rndstr, rndbuf[512], authbuf[STEL_SECRET_MAX_LEN];
	int		i, s;
	int		cipher;
	int		firstack, noptyflg, errflg, rflg;
	int		portnumber;
	extern char	*optarg;
	extern int	optind;
	struct passwd	*pwd;
	MD5_CTX		mdContext;
	unsigned char	digest[16];
	struct winsize	os_window;
	struct servent	*serv;

	void		trapper();

	char		*getstelsecret();

	/* using SYSV fashion */
	tohost = NULL;
	if (argc > 1 && argv[1][0] != '-') {
		tohost = argv[1]; /* server name */
		argv++;
		argc--;
	}

	/* parse options */
	errflg = noptyflg = rflg = 0;

	firstack = 0;
	rndstr = NULL;
	bzero(&pp, sizeof(pp));

	if ((serv = getservbyname(STELCLIENT, "tcp")) == NULL)
		portnumber = PORTNUM;
	else
		portnumber = htons(serv->s_port);
	
	while ((i = getopt(argc, argv, OPTIONS)) != -1)
		switch (i) {
			case	'l':
				strncpy(username, optarg, sizeof(username));
				strncpy(pp.username, optarg, sizeof(pp.username));
				break;

			case	't':
				noptyflg++;
				break;

			case	'n':
				if (firstack & FLG_USE_SINGLE ||
				    firstack & FLG_USE_IDEA ||
				    firstack & FLG_USE_RC4) {
					fprintf(stderr, BADOPTION, i);
					errflg++;
					break;
				}
				firstack |= FLG_NO_ENCRYPTION;
				break;

			case	'1':
				if (firstack & FLG_NO_ENCRYPTION ||
				    firstack & FLG_USE_IDEA ||
				    firstack & FLG_USE_RC4) {
					fprintf(stderr, BADOPTION, i);
					errflg++;
					break;
				}
				firstack |= FLG_USE_SINGLE;
				break;

			case	'i':
				if (firstack & FLG_NO_ENCRYPTION ||
				    firstack & FLG_USE_SINGLE ||
				    firstack & FLG_USE_RC4) {
					fprintf(stderr, BADOPTION, i);
					errflg++;
					break;
				}
				firstack |= FLG_USE_IDEA;
				break;

			case	'4':
				if (firstack & FLG_NO_ENCRYPTION ||
				    firstack & FLG_USE_SINGLE ||
				    firstack & FLG_USE_IDEA) {
					fprintf(stderr, BADOPTION, i);
					errflg++;
					break;
				}
				firstack |= FLG_USE_RC4;
				break;

			case	'm':
				if (firstack & FLG_NO_ENCRYPTION) {
					fprintf(stderr, BADOPTION, i);
					errflg++;
					break;
				}
				firstack |= FLG_SMALL_MODULUS;
				break;

			case	'M':
				if (firstack & FLG_NO_ENCRYPTION) {
					fprintf(stderr, BADOPTION, i);
					errflg++;
					break;
				}
				firstack |= FLG_LARGE_MODULUS;
				break;

			case	'a':
				firstack |= FLG_ACTIVE_ATTACKS;
				break;

			case	'r':
				if (firstack & FLG_NO_ENCRYPTION) {
					fprintf(stderr, BADOPTION, i);
					errflg++;
					break;
				}
				rflg++;
				break;

			case	'p':
				portnumber = atoi(optarg);
				if (portnumber < 0) {
					fprintf(stderr, "bad port number\n");
					errflg++;
				}
				break;

			case	'v':
				firstack |= FLG_BE_VERBOSE;
				break;

			case	'D':
				firstack |= FLG_BE_VERBOSE;
				firstack |= FLG_DEBUG;
				break;

			case	'e':
				firstack |= FLG_NO_ESCAPE;
				break;

			case	'h':
				(void)fprintf(stderr, USAGE,
						VERSION, STELCLIENT, PORTNUM);
				exit(0);
				break;
			
			case	'?':
			case	':':
				errflg++;
				break;
		}

	if (errflg) {
		(void)fprintf(stderr, "bad options -- %s -h for help.\n",
								STELCLIENT);
		exit(1);
	}

	if (rflg) {
		if (getstelpass("Enter a lenghty random string", rndbuf,
							sizeof(rndbuf))) {
			(void)fprintf(stderr, "string must be provided!\n");
			exit(1);
		}
		if (strlen(rndbuf) < 16) {
			(void)fprintf(stderr, "please enter a longer string!\n");
			exit(1);
		}
		rndstr = rndbuf;
	}

	/* set command */
	for (i = 0; optind < argc; optind++) {
		i += strlen(argv[optind]) + 1;
		if (i > sizeof(pp.command) - 2) {
			(void)fprintf(stderr, "command is too large.\n");
			exit(1);
		}
		(void)strcat(pp.command, argv[optind]);
		if (optind < argc -1)
			(void)strcat(pp.command, " ");
	}

	/* set username */
	if (pp.username[0] == '\0') {
		if ((pwd = getpwuid(getuid())) == NULL) {
			(void)fprintf(stderr, "Who are you?\n");
			exit(1);
		}
		strncpy(username, pwd->pw_name, sizeof(username));
		strncpy(pp.username, pwd->pw_name, sizeof(pp.username));
	}

	/* don't use terminals when deattached from a terminal */
	if (!isatty(0)) {
		noptyflg = 1;
		pp.mode |= FLG_DONTUSE_PTY;
	}
	/* command line option */
	else if (noptyflg)
		pp.mode |= FLG_DONTUSE_PTY;

	
	if (!noptyflg && isatty(0) && ioctl(0, TIOCGWINSZ, &os_window) >= 0) {
		pp.rows = os_window.ws_row;
		pp.cols = os_window.ws_col;
	}
	else {
		pp.rows = (-1);
		pp.cols = (-1);
	}

	/* set environment */
	if ((env = (char *)getenv("TERM")) != NULL)
		strncpy(pp.TERM, env, sizeof(pp.TERM));
	if ((env = (char *)getenv("DISPLAY")) != NULL)
		strncpy(pp.DISPLAY, env, sizeof(pp.DISPLAY));
	if ((env = (char *)getenv("LINES")) != NULL)
		strncpy(pp.LINES, env, sizeof(pp.LINES));
	if ((env = (char *)getenv("COLUMNS")) != NULL)
		strncpy(pp.COLUMNS, env, sizeof(pp.COLUMNS));
	if ((env = (char *)getenv("WINDOWID")) != NULL)
		strncpy(pp.WINDOWID, env, sizeof(pp.WINDOWID));
	/*
	if ((env = (char *)getenv("PATH")) != NULL)
		strncpy(pp.PATH, env, sizeof(pp.PATH));
	*/

	if (tohost == NULL) {
		printf("Hostname? ");
		if (fgets(hostbuf, sizeof(hostbuf), stdin) == NULL) {
			(void)fprintf(stderr, "no host?\n");
			exit(1);
		}
		tohost = hostbuf;
		rip(tohost);
		if (!strlen(tohost)) {
			(void)fprintf(stderr, "No host?\n");
			exit(1);
		}
	}

#ifdef __ultrix
	(void)openlog(STELCLIENT, LOG_WARNING);
#else
	(void)openlog(STELCLIENT, LOG_PID | LOG_ODELAY, LOG_DAEMON);
#endif

	if ((s = callsocket(tohost, portnumber)) == -1)
		exit(1);

	if (firstack & FLG_BE_VERBOSE)
		printf("Connected to %s on port %d.\n", tohost, portnumber);
	else
		printf("Connected to %s.\n", tohost);
	
#ifdef ASCIIBANNER
	/* silently get and ignore banner from server side */
	if (firstack & FLG_DEBUG)
		printf("\nEATING BANNER FROM SERVER\n");
	(void)eatbanner(s);
#endif
	
	/* tell master about user selected options */
	if (firstack & FLG_DEBUG)
		printf("\nSENDING FIRSTACK 0x%02X TO SERVER\n", firstack);
	i = htonl(firstack);
	if (write(s, &i, sizeof(firstack)) != sizeof(firstack)) {
		perror("write(firstack)");
		exit(1);
	}

	pp.mode = htonl(pp.mode);
	pp.rows = htonl(pp.rows);
	pp.cols = htonl(pp.cols);
	pp.version = htonl(VERSION);

	if (!(firstack & FLG_NO_ENCRYPTION)) {
		memcpy(pp.random, (char *)randomdigest(), 16);
		if (firstack & FLG_DEBUG) {
			hexdump("client stream randomizer: ", pp.random, 16);
		}
	}

	MD5Init(&mdContext);
	/* skip digest itself */
	MD5Update(&mdContext, &pp, sizeof(pp) - 16);
	MD5Final(pp.digest, &mdContext);

	if (!(firstack & FLG_NO_ENCRYPTION)) { /* encryption required */
		/* exchange keys */
		if (dhexchange(s, CLIENT_SIDE, firstack, rndstr) < 0) {
			perror("dhexchange()");
			exit(1);
		}

		/* use IDEA encryption */
		if (firstack & FLG_USE_IDEA) {
			encmode = "IDEA";
			cipher = IDEA_CIPHER;
		}

		/* use SINGLE DES encryption */
		else if (firstack & FLG_USE_SINGLE) {
			encmode = "SINGLE DES";
			cipher = DES_CIPHER;
		}

		/* use IDEA encryption */
		else if (firstack & FLG_USE_RC4) {
			encmode = "RC4";
			cipher = RC4_CIPHER;
		}

		/* use triple DES encryption by default */
		else {
			encmode = "TRIPLE DES";
			cipher = TRIPLEDES_CIPHER;
		}

		/* set client IV */
		memcpy(&c_IV, pp.random, 8);

		/* encrypt pp, skipping IV */
		cryptbuf((char *)&pp + 8, sizeof(pp) - 8, DES_ENCRYPT, cipher,
								CLIENT_SIDE);
	}

	if (firstack & FLG_DEBUG) {
		printf("\nSENDING ENVINFO TO SERVER\n");
		hexdump("first envinfo 20 bytes: ", &pp, 20);
		hexdump("next envinfo 20 random bytes : ",
							(char *)&pp + 20, 20);
		hexdump("next envinfo 20 random bytes : ",
							(char *)&pp + 40, 20);
	}

	if (write_data(s, &pp, sizeof(pp)) != sizeof(pp)) {
		perror("write()");
		exit(1);
	}

	/* we have send the packet to server, now let's listen to its reply */
	if ((i = read(s, &ppp, sizeof(ppp)) != sizeof(ppp))) {
		if (!i) {
			printf("Server has gone away.\n");
			exit(1);
		}
		perror("read()");
		exit(1);
	}

	if (!(firstack & FLG_NO_ENCRYPTION)) { /* encryption required */
		/* set server IV */
		memcpy(&s_IV, ppp.random, 8);

		/* decrypt it skipping server IV */
		cryptbuf((char *)&ppp + 8, sizeof(ppp) - 8, DES_DECRYPT,
							cipher, SERVER_SIDE);

	}

	/* check ppp integrity */
	MD5Init(&mdContext);
	/* skip digest itself */
	MD5Update(&mdContext, &ppp, sizeof(ppp) - 16);
	MD5Final(digest, &mdContext);

	if (memcmp(digest, ppp.digest, 16)) {
		printf("Received corrupted data from server!\n");
		exit(1);
	}

	ppp.sreply = ntohl(ppp.sreply);

	if (firstack & FLG_DEBUG) {
		printf("\nREAD REPLY FROM SERVER\n");
		printf("sreply flag: 0x%02X\n", ppp.sreply);

		if (!(firstack & FLG_NO_ENCRYPTION))
			hexdump("server stream randomizer: ", ppp.random, 16);
	}

	if (ppp.sreply & MSG_WRONG_VERSION) {
		printf("%s version mismatch.  Can not talk with server.\n",
								STELCLIENT);
		exit(1);
	}
					

	/* this is is weak, and it is intended as an experimental feature
	only.  the point is that /etc/stelsecret should be readable by the
	user.  this is no good since the user could make /etc/stelsecret
	public and thus completely defeat the mutual authentication method.
	actually, stel is supposed to be run as a SUID to root program in
	order to take advantage of this feature. I do not like SUIDed
	programs, though. */
	if (ppp.sreply & MSG_SYSTEM_AUTH_REQUIRED) {
		char	*systelsec;

		printf("Host mutual authentication is required by %s.\n",
								tohost);

		if ((systelsec = getstelsecret(ETC_STEL_SECRET_NAME, tohost)) == NULL) {
			printf("Cannot access %s\n", ETC_STEL_SECRET_NAME);
			if (getstelpass("Please enter it manually", authbuf,
							sizeof(authbuf))) {
				(void)fprintf(stderr, "No secret?\n");
				exit(1);
			}
			systelsec = authbuf;
		}

		i = mauth(s, systelsec, CLIENT_SIDE, firstack);
		bzero(systelsec, strlen(systelsec));
		if (i == -1) {
			perror("mauth()");
			exit(1);
		}
		else if (i == 1) {
			fprintf(stderr, "Host mutual authentication failed!\n");
			exit(1);
		}
		printf("Host mutual authentication OK\n");
	}

	/* restore ids, in case stel was a SUID program */
	if (setgid(getgid()) == -1) {
		perror("setgid()");
		exit(1);
	}
	if (setuid(getuid()) == -1) {
		perror("setuid()");
		exit(1);
	}

	if (ppp.sreply & MSG_LOGIN_AUTH_REQUIRED) {
		printf("User mutual authentication required by %s@%s.\n",
							username, tohost);

		if (getstelpass("Enter stelsecret", authbuf, sizeof(authbuf))) {
			(void)fprintf(stderr, "No secret?\n");
			exit(1);
		}

		i = mauth(s, authbuf, CLIENT_SIDE, firstack);
		if (i == -1) {
			perror("mauth()");
			exit(1);
		}
		else if (i == 1) {
			fprintf(stderr, "User mutual authentication failed!\n");
			exit(1);
		}
		printf("User mutual authentication OK\n");
	}
		
	if (ppp.sreply == MSG_MUST_AUTH) {
		printf(
			"Mutual authentication required -- NOT using encryption, aborting...\n");
		exit(1);
	}
	
	if (ppp.sreply == MSG_CORRUPTED_DATA) {
		printf("Server received corrupted data from us!\n");
		exit(1);
	}

	/* server allocates a pseudo terminal */
	if (!noptyflg) {
		(void)signal(SIGINT, trapper);
		(void)signal(SIGQUIT, trapper);
		(void)signal(SIGHUP, trapper);
		(void)signal(SIGTERM, trapper);

		/* set terminal to raw */
		if (tcgetattr(0, &old) == -1) {
			perror("tcgetattr()");
			exit(1);
		}
		(void)memcpy(&new, &old, sizeof(struct termios));
		new.c_cc[VTIME] = 0;
		new.c_cc[VMIN]  = 1;
		new.c_cc[VINTR] = 0;
		new.c_lflag &= ~(ECHO | ISIG | ICANON);
		new.c_oflag &= ~(ONLCR | OPOST);
#ifndef SOLARIS
#if defined(BSD) && (BSD >= 199306)
		new.c_iflag &= ~(INLCR | ICRNL | IXON | IXOFF | 
							ISTRIP | BRKINT);
#else
		new.c_iflag &= ~(INLCR | ICRNL | IXON | IXOFF | IUCLC |
							ISTRIP | BRKINT);
#endif
#else /* !SOLARIS */
		new.c_iflag &= ~(INLCR | ICRNL | IXON | IXOFF | IUCLC |
							ISTRIP | BRKINT);
#endif /* SOLARIS */
		if (tcsetattr(0, TCSANOW, &new) == -1) {
			perror("tcgetattr()");
			exit(1);
		}
		ttyset = 1;
	}

	if (firstack & FLG_NO_ENCRYPTION)
		printf(NOUSECRYPT);
	else
		printf(USECRYPT, encmode);

	if (firstack & FLG_ACTIVE_ATTACKS)
		printf(ACTIVEPROTECT);

	/* escape status */
	if (noptyflg) {
		printf(NOUSEPTY);
		printf("Escape features disabled.\r\n");
	}
	else
		printf("Escape character is '^]'.\r\n");

	cprintf("please note the PID\r\n");
	cprintf("messages without the PID are originated by remote party\r\n");
	cprintf("----GOING REMOTE----\r\n\n");

	if (makeio(0, 1, s, -1, CLIENT_SIDE, firstack, tohost,
				&old, &new, NOTIMEOUT, username)) {
		if (logfh)
			(void)close(logfh);
		if (!noptyflg)
			(void)tcsetattr(0, TCSANOW, &old);
		cprintf("Connection with %s lost.\n", tohost);
		(void)shutdown(s, 2);
		(void)close(s);
		exit(1);
	}
	if (logfh)
		(void)close(logfh);
	(void)shutdown(s, 2);
	(void)close(s);
	if (!noptyflg)
		(void)tcsetattr(0, TCSANOW, &old);
	cprintf("Connection with %s closed.\n", tohost);
	return(0);
}

void
trapper(s)
int	s;
{
	(void)signal(s, trapper);

	(void)fprintf(stderr, "%s: trapped signal %d!\n", STELCLIENT, s);
	if (ttyset)
		(void)tcsetattr(0, TCSANOW, &old);
	exit(1);
}

getstelpass(prompt, buf, len)
char	*prompt, *buf;
int	len;
{
	void		(*oldsig) (); 
	void		gp_restore_ttymode();

	if (isatty(0)) {
		printf("%s: ", prompt);
		(void)fflush(stdout);
		/* ECHO OFF */
		tcgetattr(0, &saved_ttymode);
		if ((oldsig = signal(SIGINT, SIG_IGN)) != SIG_IGN)
			(void)signal(SIGINT, gp_restore_ttymode);
		tcgetattr(0, &noecho_ttymode);
		noecho_ttymode.c_lflag &= ~ECHO;
		tcsetattr(0, TCSANOW, &noecho_ttymode);
	}
	else {
		printf("%s (echo is on): ", prompt);
		(void)fflush(stdout);
	}

	/* get password */
	if (fgets(buf, len, stdin) == NULL)
		return(1);
	rip(buf);

	if (isatty(0)) {
		/* ECHO ON */
		tcsetattr(0, TCSANOW, &saved_ttymode);
		if (oldsig != SIG_IGN)
			(void)signal(SIGINT, oldsig);
		printf("\n");
	}
	return(0);
}

/* restore - restore terminal modes when user aborts command */
void gp_restore_ttymode()
{
	if (tcsetattr(0, TCSANOW, &saved_ttymode) == -1)
		perror("tcsetattr()");
	exit(1);
}

